pkg/controller/release/release_controller.go

```diff
@@ -392,18 +433,42 @@ func (c *Controller) executeStrategyOnClusters(
 	}
 
 	var strategy *shipper.RolloutStrategy
-	var targetStep int32
+	var virtualStrategy *shipper.RolloutVirtualStrategy
 	// A head release uses it's local spec-defined strategy, any other release
 	// follows it's successor state, therefore looking into the forecoming spec.
 	if isHead {
 		strategy = rel.Spec.Environment.Strategy
-		targetStep = rel.Spec.TargetStep
+		virtualStrategy = rel.Spec.VirtualStrategy
 	} else {
 		strategy = succ.Spec.Environment.Strategy
-		targetStep = succ.Spec.TargetStep
+		virtualStrategy = succ.Spec.VirtualStrategy
+	}
+
+	progressing := releaseutil.IsReleaseProgressing(rel.Status.AchievedStep, rel.Spec.TargetStep)
+	hasAchievedTargetStep := releaseutil.HasReleaseAchievedTargetStep(rel)
+	stepsFromAchievedStep := releaseutil.StepsFromAchievedStep(rel.Status.AchievedStep, rel.Spec.TargetStep)
+	if !isHead {
+		// head is the one that's supposed to decide if this is a rolling forward or backward.
+		progressing = releaseutil.IsReleaseProgressing(succ.Status.AchievedStep, succ.Spec.TargetStep)
+		hasAchievedTargetStep = releaseutil.HasReleaseAchievedTargetStep(succ)
+		stepsFromAchievedStep = releaseutil.StepsFromAchievedStep(succ.Status.AchievedStep, succ.Spec.TargetStep)
+	}
+
+	targetStep := releaseutil.GetTargetStep(rel, succ, strategy, stepsFromAchievedStep, progressing, hasAchievedTargetStep)
+	if targetStep >= int32(len(strategy.Steps)) {
+		return rel, shippererrors.NewUnrecoverableError(
+			fmt.Errorf("no step %d in strategy", targetStep))
+	}
+
+	if virtualStrategy == nil {
+		// wait for virtual strategy to be built
+		err := fmt.Errorf("release %q missing a virtual strategy", objectutil.MetaKey(rel))
+		return rel, shippererrors.NewRecoverableError(err)
```

Comment:
This should be an unrecoverable error, as the virtual strategy eventually appearing will trigger a new event. Otherwise, shipper will keep on spinning on this until the virtual strategy appears, if ever.